pkg/plugin/object_store.go

```diff
@@ -0,0 +1,254 @@
+package plugin
+
+import (
+	"io"
+	"time"
+
+	"github.com/hashicorp/go-plugin"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"github.com/heptio/ark/pkg/cloudprovider"
+	proto "github.com/heptio/ark/pkg/plugin/generated"
+)
+
+// ObjectStorePlugin dispenses gRPC servers/clients for the ObjectStore
+// plugin.
+type ObjectStorePlugin struct {
+	plugin.NetRPCUnsupportedPlugin
+
+	impl cloudprovider.ObjectStore
+}
+
+func NewObjectStorePlugin(objectStore cloudprovider.ObjectStore) *ObjectStorePlugin {
+	return &ObjectStorePlugin{
+		impl: objectStore,
+	}
+}
+
+func (p *ObjectStorePlugin) GRPCServer(s *grpc.Server) error {
+	proto.RegisterObjectStoreServer(s, &ObjectStoreGRPCServer{impl: p.impl})
+	return nil
+}
+
+func (p *ObjectStorePlugin) GRPCClient(c *grpc.ClientConn) (interface{}, error) {
+	return &ObjectStoreGRPCClient{grpcClient: proto.NewObjectStoreClient(c)}, nil
+}
+
+// ObjectStoreGRPCClient implements the cloudprovider.ObjectStore interface and forwards
+// calls to a gRPC client.
+type ObjectStoreGRPCClient struct {
+	grpcClient proto.ObjectStoreClient
+}
+
+// Init prepares the ObjectStore for usage using the provided map of
+// configuration key-value pairs. It returns an error if the ObjectStore
+// cannot be initialized from the provided config.
+func (c *ObjectStoreGRPCClient) Init(config map[string]string) error {
+	_, err := c.grpcClient.Init(context.Background(), &proto.InitRequest{Config: config})
+
+	return err
+}
+
+// PutObject creates a new object using the data in body within the specified
+// object storage bucket with the given key.
+func (c *ObjectStoreGRPCClient) PutObject(bucket, key string, body io.Reader) error {
+	stream, err := c.grpcClient.PutObject(context.Background())
+	if err != nil {
+		return err
+	}
+
+	// read from the provider io.Reader into chunks, and send each one over
+	// the gRPC stream
+	chunk := make([]byte, 1024)
+	for {
+		n, err := body.Read(chunk)
+		if err == io.EOF {
+			_, resErr := stream.CloseAndRecv()
+			return resErr
+		} else if err != nil {
+			stream.CloseSend()
+			return err
+		}
+
+		if err := stream.Send(&proto.PutObjectRequest{Bucket: bucket, Key: key, Body: chunk[0:n]}); err != nil {
+			return err
+		}
+	}
+}
+
+// GetObject retrieves the object with the given key from the specified
+// bucket in object storage.
+func (c *ObjectStoreGRPCClient) GetObject(bucket, key string) (io.ReadCloser, error) {
+	stream, err := c.grpcClient.GetObject(context.Background(), &proto.GetObjectRequest{Bucket: bucket, Key: key})
+	if err != nil {
+		return nil, err
+	}
+
+	receive := func() ([]byte, error) {
```

Comment:
I'm guessing if we have other uses for StreamReadCloser, they're probably going to look identical to this? Consider having StreamReadCloser take in an interface?
```go
type RecvCloser interface {
  Recv() ([]byte, error)
  CloseSend() error
}
```

If we need an adapter to turn `*proto.Bytes` into `[]byte`, we could make one instead of doing the inline function. This may be premature optimization, though.